src/Symfony/Component/Form/ChoiceList/Loader/CallbackChoiceLoader.php

```diff
@@ -0,0 +1,87 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Form\ChoiceList\Loader;
+
+use Symfony\Component\Form\ChoiceList\ArrayChoiceList;
+
+/**
+ * Loads a {@link ArrayChoiceList} interface from any callable.
+ *
+ * @author Jules Pietri <jules@heahprod.com>
+ */
+class CallbackChoiceLoader implements ChoiceLoaderInterface
+{
+    /**
+     * The callable used to load the choices.
+     *
+     * @var callable
+     */
+    private $callback;
+
+    /**
+     * The loaded choice list.
+     *
+     * @var ArrayChoiceList
+     */
+    private $choiceList;
+
+    public function __construct(callable $callback)
+    {
+        $this->callback = $callback;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadChoiceList($value = null)
+    {
+        if (null !== $this->choiceList) {
+            return $this->choiceList;
+        }
+
+        return $this->choiceList = new ArrayChoiceList(call_user_func($this->callback), $value);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function loadChoicesForValues(array $values, $value = null)
+    {
+        // Optimize
+        if (empty($values)) {
+            return array();
+        }
+
+        if (null === $this->choiceList) {
+            $this->loadChoiceList($value);
```

Comment:
In fact, I realised something we can optimize those two methods a lot. For all these types using ICU libraries, we only need to load them to get the proper localized labels, because unless `choice_value` is explicitly set, the choices are the same as the values.

Hence, when `$value` is `null` we just need to return the values!
And same goes for `loadValuesForChoices`, we just need to return the choices.

I will update one final time this PR, cool :)